### DISCUSSION TITLE
Applying Group flags to an Enum

### DISCUSSION THREAD

**Discussion by aadidion** [dotnet/csharplang#5519] [Language Ideas]
Given an Enum like this:
        public enum PersonTypeEnum
        {
            Patient = 1,
            Subscriber = 2,
            OtherSubscriber = 3,
            ResponsibleParty = 4,
            Submitter = 5,
            Receiver = 6

        }

Devise a way to apply flags to the Enum items so they can be subgrouped and pulled out in code to create a sub-Enum on the fly.

Example: 
        public enum PersonTypeEnum
        {
            Patient = 1 | {Animal, Human, Other},
            Subscriber = 2 | {Animal, Human},
            OtherSubscriber = 3 | {Human},
            ResponsibleParty = 4, | {Human},
            Submitter = 5, | {Human},
            Receiver = 6 | {Payer}
        }
 
var HumanTypeEnum = PersonTypeEnum.Where(m => m.Human)

This construct saves having to code 5 different Enums and with filters to create subgroups I can iterate through the ones I need based on a specific criteria while coding.  

Also, I believe this is a new construct.  I don't believe anybody else has it.

**Comment by TheUnlocked** [comment]
Edit: Having read HaloFour's comment, it's possible I misinterpreted the suggestion. Is the idea that a patient would be a human, animal, or other, or that it would be a human, animal, AND other?

Enums are just numbers. If I tried to convert `1` to a `PersonTypeEnum`, what subgroup would it have? And how would the subgroup of a value be represented in memory?

It's possible that DUs may be closer to what you want, whenever those get designed and shipped.

**Comment by HaloFour** [comment]
Something like this could be implemented via extension methods:

```cs
public static class PersonTypeEnumExtensions {
    public static bool IsHuman(this PersonTypeEnum e) {
        return e == PersonTypeEnum.Patient
            || e == PersonTypeEnum.Subscriber
            || e == PersonTypeEnum.OtherSubscriber
            || e == PersonTypeEnum.ResponsibleParty
            || e == PersonTypeEnum.Submitter;
    }
    
    public static bool IsAnimal(this PersonTypeEnum e) {
        return e == PersonTypeEnum.Patient
            || e == PersonTypeEnum.Subscriber;
    }
    
    public static bool IsOther(this PersonTypeEnum e) {
        return e == PersonTypeEnum.Patient
            || e == PersonTypeEnum.Receiver;
    }
}
```

You could then use them in a LINQ query like this:

```cs
var humanTypes = Enum.GetValues<PersonTypeEnum>()
    .Where(e => e.IsHuman())
    .ToArray();
```

**Reply by svick** [reply]
Or you could use attributes to mark the `enum` members:

```c#
public enum PersonTypeEnum
{
    [Animal, Human, Other]
    Patient = 1,
    [Animal, Human]
    Subscriber = 2,
    [Human]
    OtherSubscriber = 3,
    [Human]
    ResponsibleParty = 4,
    [Human]
    Submitter = 5,
    [Payer]
    Receiver = 6
}
```

Then use reflection to figure out which member is marked with which attribute (probably combined with caching to make it more efficient).
